precompiles/erc20/integration_test.go

```diff
@@ -0,0 +1,787 @@
+			Expect(err).ToNot(HaveOccurred(), "failed to fund account")
+
+			senderBalancePre, err := s.grpcHandler.GetBalance(sender.AccAddr, s.tokenDenom)
+			Expect(err).ToNot(HaveOccurred(), "failed to get balance")
+			Expect(senderBalancePre.Balance.Amount.Int64()).To(Equal(fundAmountSender.Int64()), "expected different balance before transfer")
+
+			receiverBalancePre, err := s.grpcHandler.GetBalance(receiver.AccAddr, s.tokenDenom)
+			Expect(err).ToNot(HaveOccurred(), "failed to get balance")
+			Expect(receiverBalancePre.Balance.Amount.Int64()).To(Equal(fundAmountReceiver.Int64()), "expected different balance before transfer")
+
+			// Transfer tokens
+			txArgs, transferArgs := s.getTxAndCallArgs(callType, contractAddr)
+			transferArgs.MethodName = erc20.TransferMethod
+			transferArgs.Args = []interface{}{receiver.Addr, amount}
+
+			transferCheck := passCheck.WithExpEvents(erc20.EventTypeTransfer)
+
+			_, _, err = s.callContractAndCheckLogs(sender.Priv, txArgs, transferArgs, transferCheck)
+			Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+			senderBalancePost, err := s.grpcHandler.GetBalance(sender.AccAddr, s.tokenDenom)
+			Expect(err).ToNot(HaveOccurred(), "failed to get balance")
+			Expect(senderBalancePost.Balance.Amount.Int64()).To(Equal(senderBalancePre.Balance.Amount.Int64()-amount.Int64()), "expected different balance after transfer")
+
+			receiverBalancePost, err := s.grpcHandler.GetBalance(receiver.AccAddr, s.tokenDenom)
+			Expect(err).ToNot(HaveOccurred(), "failed to get balance")
+			Expect(receiverBalancePost.Balance.Amount.Int64()).To(Equal(receiverBalancePre.Balance.Amount.Int64()+amount.Int64()), "expected different balance after transfer")
+		},
+			Entry(" - direct call", directCall),
+			// NOTE: we are not passing the contract call here because that requires an authorization which is
+			// a separate test case.
+		)
+
+		// TODO: is this the behavior we want? Makes sense right because the contract is not a wallet?
+		DescribeTable("it should return an error trying to call from a smart contract", func(callType int) {
+			receiver := s.keyring.GetAddr(1)
+			fundAmount := big.NewInt(300)
+			amount := big.NewInt(100)
+
+			// Fund account with some tokens
+			err = s.network.FundAccount(sender.AccAddr, sdk.Coins{
+				{Denom: s.tokenDenom, Amount: sdk.NewIntFromBigInt(fundAmount)},
+			})
+			Expect(err).ToNot(HaveOccurred(), "failed to fund account")
+
+			// Transfer tokens
+			txArgs, transferArgs := s.getTxAndCallArgs(callType, contractAddr)
+			transferArgs.MethodName = erc20.TransferMethod
+			transferArgs.Args = []interface{}{receiver, amount}
+
+			_, _, err = s.callContractAndCheckLogs(sender.Priv, txArgs, transferArgs, execRevertedCheck)
+			Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+		},
+			// NOTE: we are not passing the direct call here because this test is specific to the contract calls
+			Entry(" - through contract", contractCall),
+		)
```

Comment:
Because we are not passing and checking the origin in the ERC20 transactions right now, it's not possible to call the transactions using a smart contract right now? I would assume we want to enable using smart contracts too right?